### PR TITLE
New server

### PR DESCRIPTION
**Pull request by yjhn**
Added:
new server based on ASP .NET Core
new server Api client

### CODE REVIEW

File: Backend/FileReader.cs [hunk 1]
```diff
@@ -79,64 +79,126 @@ internal User GetUserData(string username)
             return null;
         }
 
-        internal Task<List<Car>> GetAllCarData()
+        //internal Task<List<Car>> GetAllCarData()
```
Comment (MartynasRaf) [inline Backend/FileReader.cs]: Maybe these lines should be deleted as commented code can be reached through version control :)

File: Backend/IDatabase.cs [hunk 2]
```diff
@@ -4,37 +4,46 @@ namespace Backend
...
+        public Car GetCar(int id);
         // args: car serialized to UTF-8 JSON
-        public bool AddCarJson(byte[] car);
+        public bool AddCar(string username, string password, Car car);
 
+        public bool UpdateCar(string username, string password, Car car);
         // returns: true, if removal is successful, false if not
-        public bool DeleteCar(int id);
-
+        public bool DeleteCar(int carId, string username, string password);
```
Comment (MartynasRaf) [inline Backend/IDatabase.cs]: Password should never appear as a string in code. Check the user token instead. https://en.wikipedia.org/wiki/Access_token

File: DataTypes/Car.cs [hunk 3]
```diff
@@ -87,6 +87,11 @@ public bool Equals(Car other)
 
         // required for serialization to work
         public Car() { }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as Car);
```
Comment (MartynasRaf) [inline DataTypes/Car.cs:93]: Shouldn't equal method compare all relevent parameters and the return true/false based on that?

File: Frontend/IApiWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Frontend
+{
+    public interface IApiWrapper
+    {
+        /*
+         * int response codes:
```
Comment (MartynasRaf) [inline Frontend/IApiWrapper.cs:11]: Maybe enums would make the code clearer if used for response codes

File: ServerV2/Controllers/WeatherForecastController.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace ServerV2.Controllers
+{
+    [ApiController]
+    [Route("[controller]")]
+    public class WeatherForecastController : ControllerBase
+    {
+        private static readonly string[] Summaries = new[]
+        {
+            "Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"
```
Comment (MartynasRaf) [inline ServerV2/Controllers/WeatherForecastController.cs]: Is this code relevant to the project?

File: ServerV2/WeatherForecast.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+using System;
+
+namespace ServerV2
+{
+    public class WeatherForecast
+    {
```
Comment (MartynasRaf) [inline ServerV2/WeatherForecast.cs]: Relevance?

File: DataTypes/Car.cs (hunk 3 above)
Comment (yjhn) [reply DataTypes/Car.cs:93]: Car is identified by its ID, so I thought that since no two different cars can have the same ID, I should not waste time comparing other properties

File: ServerV2/Controllers/WeatherForecastController.cs (hunk 5 above)
Comment (yjhn) [reply ServerV2/Controllers/WeatherForecastController.cs]: Forgot to delete

File: DataTypes/Car.cs (hunk 3 above)
Comment (somkamarius) [reply DataTypes/Car.cs:93]: I think it makes sense to compare by ID only, atleast in this case. Seems like we won't really compare vehicles based on their paremeters just yet (or atleast won't override Equals for this cause)

Comment (MartynasRaf) [reply DataTypes/Car.cs:93]: What if there's an error where two vehicles have the same ID, but different params?

Comment (yjhn) [reply DataTypes/Car.cs:93]: If this happens, server should crash (at least in my opinion), because it isn't working correctly

File: Backend/IDatabase.cs (hunk 2 above)
Comment (yjhn) [reply Backend/IDatabase.cs]: https://en.wikipedia.org/wiki/Representational_state_transfer
Quoting from there: "...using a uniform and predefined set of stateless operations"
At least as far as I know REST APIs should not store temporary state (such as access tokens).
Unless we are not making REST API

File: DataTypes/Car.cs (hunk 3 above)
Comment (MartynasRaf) [reply DataTypes/Car.cs:93]: Under no circumstances should the server crash. It should throw an exception something like DuplicateIDException and handle this behavior to avoid crashing.

File: Backend/IDatabase.cs (hunk 2 above)
Comment (MartynasRaf) [reply Backend/IDatabase.cs]: image

https://www.oauth.com/oauth2-servers/access-tokens/#:~:text=Access%20tokens%20are%20the%20thing,in%20transit%20and%20in%20storage.

Comment (yjhn) [reply Backend/IDatabase.cs]: Our current implementation is keeping user credentials private.
At least I see no big problem with current implementation.

Comment (yjhn) [reply Backend/IDatabase.cs]: Will use HTTP Basic authentication (at least for now)
